TYP: add type annotations to a couple utility functions in nddata

neutrinoceros opened the pull request:
### Description
A by-product of my reviewing #16544

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

eerovaher commented on astropy/nddata/utils.py:
```diff
@@ -25,6 +28,16 @@
+    from typing_extensions import TypeVar
+
+    T = TypeVar("T")
+    SingleOrTuple: TypeAlias = T | tuple[T, ...]
```
This [`TypeAlias`](https://docs.python.org/3/library/typing.html#typing.TypeAlias) is causing problems for Sphinx. Luckily `int | tuple[int, ...]` or `float | tuple[float, ...]` are so simple that we don't really need an alias with a [`TypeVar`](https://docs.python.org/3/library/typing.html#typing.TypeVar) for them.

eerovaher commented on astropy/nddata/utils.py:
```diff
@@ -25,6 +28,16 @@
+
+    T = TypeVar("T")
+    SingleOrTuple: TypeAlias = T | tuple[T, ...]
+    ModeT: TypeAlias = Literal["partial", "trim", "strict"]
```
Do you think it would be a good idea to use [`Enum`](https://docs.python.org/3.10/library/enum.html#creating-an-enum) for this? (in a separate pull request)

neutrinoceros replied:
Internally yes, but it shouldn't affect the user visible interface, hence the type hint should be stable. I'm curating a list of such Enum opportunities, I'll add it there.

eerovaher replied:
In this pull request `ModeT` is causing the documentation build to fail because Sphinx is trying to add a link to it in the function signatures where it is used as an annotation, but it can't find its definition. Any type aliases defined in a `if TYPE_CHECKING:` block and used in public functions will cause such failures. If the modes are not listed in an `Enum` yet then it would be best to write them out in each signature that needs them.

neutrinoceros replied:
damn I somehow missed that comment...

eerovaher commented on astropy/nddata/utils.py:
```diff
@@ -25,6 +28,16 @@
 ]
 
+if TYPE_CHECKING:
+    from numbers import Number
```
What's the reason for using [`Number`](https://docs.python.org/3/library/numbers.html#numbers.Number) here? There's a few places in `units` where [`Fraction`](https://docs.python.org/3/library/fractions.html#fractions.Fraction) is a valid input type, so we use [`Real`](https://docs.python.org/3/library/numbers.html#numbers.Real) as a type annotation because of that, but is [`float`](https://docs.python.org/3/library/functions.html#float) not good enough in this file?

neutrinoceros replied:
I used `Number` in places where I assumed complex values could be valid, but if I'm wrong then `float` would be perfectly fine instead.

eerovaher replied:
Even though `float` is not a subclass of `complex`, [the `complex` type does encompass `float`](https://peps.python.org/pep-0484/#the-numeric-tower). If the annotations should explicitly allow `numpy` datatypes then consider using [`numbers.Complex`](https://docs.python.org/3/library/numbers.html#numbers.Complex) instead.

eerovaher commented on astropy/nddata/utils.py:
```diff
@@ -264,7 +304,9 @@ def extract_array(
 
-def add_array(array_large, array_small, position):
+def add_array(
+    array_large: np.ndarray, array_small: np.ndarray, position: tuple[float]
```
Should this be
```suggestion
    array_large: np.ndarray, array_small: np.ndarray, position: tuple[float, ...]
```

neutrinoceros replied:
yes, thanks !

eerovaher reviewed (commented):
Maybe it would be a good idea to introduce a type variable for the datatype of the arrays, like [`astropy/utils/shapes.py`](https://github.com/astropy/astropy/blob/main/astropy/utils/shapes.py) does.

neutrinoceros commented:
Thanks for your feedback @eerovaher. I should have addressed all your comments

eerovaher commented on astropy/nddata/utils.py:
```diff
@@ -25,6 +28,21 @@
+
+    from numpy.typing import NDArray
+
+    # These could be generalized as
+    # T = TypeVar("T")
+    # SingleOrTuple: TypeAlias = T | tuple[T, ...]
+    # but sphinx chokes on this definition at the time of writing
+    SingleOrTupleInt: TypeAlias = int | tuple[int, ...]
+    SingleOrTupleFloat: TypeAlias = float | tuple[float, ...]
```
As far as Sphinx is concerned defining `SingleOrTupleInt` and `SingleOrTupleFloat` is not any better than `SingleOrTuple` because these definitions are still hidden from Sphinx in an `if TYPE_CHECKING:` block. You need to write out `int | tuple[int, ...]` and `float | tuple[float, ...]` in the type signatures.

neutrinoceros replied:
or maybe I can move them out of the `if TYPE_CHECKING` block instead ? (maybe even bring back my `SingleOrTuple` generic ??)

eerovaher replied:
I think you'd have to make `SingleOrTuple` a part of the public API or else Sphinx cannot link to it. I don't think it's worth it. `int | tuple[int, ...]` and `float | tuple[float, ...]` are a simple enough that we don't need aliases for them.

eerovaher commented on astropy/nddata/utils.py:
```diff
@@ -152,6 +169,28 @@ def overlap_slices(large_array_shape, small_array_shape, position, mode="partial
+    shape: int | tuple[int, ...],
+    position: float | tuple[float, ...],
+    mode: Literal["partial", "trim", "strict"],
+    fill_value: Complex,
```
I'm not familiar enough with `nddata` to know if complex numbers are relevant here. If they are then this is annotation is correct. If they are not then we should use `numbers.Real` instead of `numbers.Complex`.

neutrinoceros replied:
Yeah I used Complex just to be on the safe side and stay as general as possible but I'd like subpackage maintainers to weigh in here specifically.

eerovaher reviewed (commented):
I don't have any typing suggestions left, other than the one remark that requires more familiarity with `nddata` than I have. Sub-package maintainers should have the final say.

neutrinoceros commented:
rebased to fix a trivial conflict

neutrinoceros commented:
rebased again to refresh CI and fix conflicts. This shouldn't block 7.0 (obviously), so I'm going to move it to 7.1

neutrinoceros commented:
sorry about the delay in my reply. I've updated the PR as suggested by @mwcraig but also replaced `numbers` protocol with simple `float` and `complex` annotations, which I find better suited in situations where numpy types are accepted. This should be ready for final review now 🤞🏻

mwcraig reviewed (approved):
Thanks @neutrinoceros!
